lib/resource-timing.js

```diff
@@ -0,0 +1,228 @@
+   */
+  if(!Date.now) {
+    Date.now = function now() {
+      return new Date().getTime();
+    };
+  }
+
+  var defaults = function(o, d) {
+      for(var prop in d) {
+        if(!o.hasOwnProperty(prop)) {
+          o[prop] = d[prop];
+        }
+      }
+    },
+    contains = function(array, value) {
+      if(Array.prototype.indexOf) {
+        return array.indexOf(value) != -1;
+      } else {
+        var i, length = array.length;
+        for(i = 0; i < length; i++) {
+          if(array[i] === value) {
+            return true;
+          }
+        }
+        return false;
+      }
+    };
+
+  /**
+   * Frontend Web Performance Data Gathering
+   *
+   * @class SurfNPerf
+   */
+  var SurfNPerfRT = function() {
+      this._data = {
+        custom: {},
+        marks: {},
+        highResMarks: {},
+        events: {}
+      };
+
+      this._navigationTiming = null;
+      this._highResTime = null;
+      this._userTiming = null;
+
+      this._navigationTimingEvents = {
+        a: ["navigationStart", "unloadEventEnd", "unloadEventStart", "redirectStart", "redirectEnd", "fetchStart", "domainLookupStart", "domainLookupEnd", "connectStart", "secureConnectionStart", "connectEnd", "requestStart", "responseStart", "responseEnd", "domLoading"],
+        b: ["domInteractive", "domContentLoadedEventStart", "domContentLoadedEventEnd", "domComplete", "loadEventStart", "loadEventEnd"]
+      };
+
+      this.initialize();
+    },
+    SNPRTProto = SurfNPerfRT.prototype;
+
+  SNPRTProto._inList = function(origin, options) {
+    if (options.hasOwnProperty("whitelist")) {
+      if (options.whitelist.indexOf(origin) === -1) {
+        return false;
+      } else {
+        return true;
+      }
+    } else if (options.hasOwnProperty("blacklist")) {
+      if (options.blacklist.indexOf(origin) === -1) {
+        return true;
+      } else {
+        return false;
+      }
+    }
+  }
+
+  SNPRTProto.getOrigins = function(options) {
+    if(window.performance.getEntriesByType) {
+      var resources = window.performance.getEntriesByType("resource");
+      var origins = [];
+      for var resource in resources {
+        var url = new URL(resources[resource].name);
+        if (origins.indexOf(url.origin) === -1 && this._inList(url.origin, options)) {
+          origins.push(url.origin);
+        }
+      }
+
+      return origins;
+    } else {
+      return null;
+    }
+  };
+
+  SNPRTProto.getResourcesFromOrigin = function(origin) {
+    if(window.performance.getEntriesByType) {
+      var resources = window.performance.getEntriesByType("resource");
+      var resourcesFromOrigin = [];
+      for var resource in resources {
+        var url = new URL(resources[resource].name);
+        if (url.origin === origin) {
+          resourcesFromOrigin.push(resources[resource]);
+        }
+      }
+      return resourcesFromOrigin;
+    } else {
+      return null;
+    }
+  };
+
+  SNPRTProto.isBufferFull = function() {
```

Comment:
for this to work, you're going to have to add this logic to the code in the `<head>`:
```js
  if(window.performance) {
    window.performance.onresourcetimingbufferfull = function(){
      SURF_N_PERF.resourcetimingbufferfull = true;
    }
  }
```
and then this isBufferFull's function's logic becomes just:
```js
    if(this._resourceTiming) {
      return !!SURF_N_PERF.resourcetimingbufferfull;
    } else {
      return null;
    }
```